src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/DistributedTM/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,296 @@
+                logger.LogTrace($"{partition} Storing v{this.key.CommittedSequenceId} {batchOperation.Count} table ops");
+
+            await StoreBatch(batchOperation).ConfigureAwait(false);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+                logger.LogTrace($"{partition} Stored v{this.key.CommittedSequenceId} eTag={key.ETag}");
+
+            return key.ETag;
+        }
+
+        private bool FindState(long sequenceId, out int pos)
+        {
+            pos = 0;
+            while (pos < states.Count)
+            {
+                switch (states[pos].Key.CompareTo(sequenceId))
+                {
+                    case 0:
+                        return true;
+                    case -1:
+                        pos++;
+                        continue;
+                    case 1:
+                        break;
+                }
+            }
+            return false;
+        }
+
+        private async Task<KeyEntity> ReadKey()
+        {
+            var query = new TableQuery<KeyEntity>()
+                .Where(AzureStorageUtils.PointQuery(this.partition, KeyEntity.RK));
+            TableQuerySegment<KeyEntity> queryResult = await table.ExecuteQuerySegmentedAsync(query, null).ConfigureAwait(false);
+            return queryResult.Results.Count == 0
+                ? new KeyEntity() { PartitionKey = this.partition }
+                : queryResult.Results[0];
+        }
+
+        private async Task<List<KeyValuePair<long, StateEntity>>> ReadStates()
+        {
+            var query = new TableQuery<StateEntity>()
+                .Where(AzureStorageUtils.RangeQuery(this.partition, StateEntity.RK_MIN, StateEntity.RK_MAX));
+            TableContinuationToken continuationToken = null;
+            var results = new List<KeyValuePair<long, StateEntity>>();
+            do
+            {
+                TableQuerySegment<StateEntity> queryResult = await table.ExecuteQuerySegmentedAsync(query, continuationToken).ConfigureAwait(false);
+                foreach (var x in queryResult.Results)
+                {
+                    results.Add(new KeyValuePair<long, StateEntity>(x.SequenceId, x));
+                };
+                continuationToken = queryResult.ContinuationToken;
+            } while (continuationToken != null);
+            return results;
+        }
+
+        private async Task StoreBatch(TableBatchOperation batchOperation)
+        {
+            try
+            {
+                if (batchOperation.Count < AzureTableConstants.MaxBatchSize)
+                {
+                    await table.ExecuteBatchAsync(batchOperation).ConfigureAwait(false);
+                }
+                else
+                {
+                    var portion = new TableBatchOperation();
+                    var enumerator = batchOperation.GetEnumerator();
+                    while (enumerator.MoveNext())
+                    {
+                        portion.Add(enumerator.Current);
```

Comment:
nifty.   I didn't know you could move operations from batch to batch! :)

Reply:
Yeah. But I'll have to rewrite this part to make it easier to add the key to each batch.